Gui: New and enhanced color selection logic for Expression Input Dialog

rmTizi opened the pull request:
As discussed on #170

In summary, the Expression Input dialog now respects user color choices from Output Window preferences while providing properties for style sheets to add explicit support for the dialog states.

It also handles existing unaware style sheets by using readable default colors both in light and dark mode themes accordingly.

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [x]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [x]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [x]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [x]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [x]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the [FreeCAD bug tracker issue number](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) in case a particular commit solves or is related to an existing issue on the tracker. Ex: `Draft: fix typos - fixes #0004805`

Base unit test ran without error, Visual inspection of all the existing themes to confirm effects as expected.  

Unfortunately only able to test on Windows

rmTizi committed:
Gui: New and enhanced color selection logic for Expression Input Dialog

As discussed on #170

In summary, the Expression Input dialog now respects user color choices from Output Window preferences while providing properties for style sheets to add explicit support for the dialog states.

It also handles existing unaware style sheets by using readable default colors both in light and dark mode themes accordingly.

rmTizi committed:
Merge branch 'realthunder:LinkDaily' into LinkDaily

realthunder commented on src/Gui/DlgExpressionInput.h:
```diff
@@ -69,7 +102,16 @@ class GuiExport DlgExpressionInput : public QDialog
     void show();
-    void applyStylesheet();
+
+public Q_SIGNAL:
```
This looks wrong. You should use `Q_SIGNALS` to mark the following group of functions as signals. `Q_SIGNAL` only marks the current line as a signal function.

On the other hand, I don't think you want signals here. What you need is a `WRITE` accessor, so that you'll know when someone has written to the property. `NOTIFY` is really for notifying other code that uses `DlgExpressionInput`.

rmTizi replied:
I replaced the signals by accessor functions, it does double the number of functions to add in the header and implement, but it is indeed a better solution.

realthunder commented on src/Gui/DlgExpressionInput.cpp:
```diff
@@ -633,4 +581,135 @@ void ProxyWidget::paintEvent(QPaintEvent *)
+                            &this->_errorColor, &this->_errorBackgroundColor);
+
+    //Old logic to decide the border and background colors of the checkboxes if there no system background is on
+    //I failed to understand excatly *why* this was needed so for now I only removed redundant variables
```
The reason of having `noBackground` is that on some Linux system, transparent top level pop up window does not work, resulting black background instead of transparent. So the user is offer a choice to turn off `noBackground` and change to a normal dialog, in which case the checkbox style shall follow the main stylesheet.

rmTizi replied:
I see, I did read the bit about the mouse events but somehow the initial description of noBackground eluded me, I modified the comment to reflect the understanding.

rmTizi committed:
Gui: Replaced the use of Q_SIGNAL in Expression Input Dialog and implemented Accessors instead

Per pull request review notes, also updated comments to reflect understanding of the check box styling needs

rmTizi committed:
Merge branch 'realthunder:LinkDaily' into LinkDaily
